library/item/item_group/src/main/java/org/quiltmc/qsl/item/group/impl/ItemGroupIconRendererRegistry.java

```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright 2022 QuiltMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.quiltmc.qsl.item.group.impl;
+
+import java.util.Map;
+
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.Nullable;
+import org.quiltmc.qsl.item.group.api.client.ItemGroupIconRenderer;
+
+import it.unimi.dsi.fastutil.objects.Object2ObjectOpenHashMap;
+import net.fabricmc.api.EnvType;
+import net.fabricmc.api.Environment;
+import net.minecraft.item.ItemGroup;
+
+@ApiStatus.Internal
+@Environment(EnvType.CLIENT)
+public final class ItemGroupIconRendererRegistry {
+	private static final Map<ItemGroup, ItemGroupIconRenderer<? extends ItemGroup>> RENDERERS = new Object2ObjectOpenHashMap<>();
+
+	@SuppressWarnings("unchecked")
+	public static <IG extends ItemGroup> ItemGroupIconRenderer<IG> register(IG itemGroup, ItemGroupIconRenderer<IG> renderer) {
+		return (ItemGroupIconRenderer<IG>) RENDERERS.put(itemGroup, (ItemGroupIconRenderer<? extends ItemGroup>) renderer);
```

Comment:
```java
(ItemGroupIconRenderer<? extends ItemGroup>) renderer
```
This cast isn't necessary

Reply:
Actually it is, it won't compile without it on my IDE

Reply:
If that's the case, that seems like a bug in your IDE. Since `IG` extends `ItemGroup`, a parameter of `IG` is already a `? extends ItemGroup`. However, no harm in leaving it in, I guess, though it's a bit confusing to read.

Reply:
wait hm, it does compile now for some reason, my IDE is haunted I guess perfect for spooky month!